File: cli/src/docker-conflicts.ts
import { spawnSync } from 'child_process';
import { LABEL_PROJECT } from '@/labels.js';
import { plannedComposeNames } from '@/generator.js';
import { projectId } from '@/labels.js';
import type { DevcontainerConfig } from '@/types.js';

export interface Conflict {
  kind: 'container' | 'network';
  name: string;
  owner: string;
}

function dockerAvailable(): boolean {
  const r = spawnSync('docker', ['version', '--format', '{{.Client.Version}}'], {
    encoding: 'utf8',
  });
  return r.status === 0;
}

function listExisting(kind: 'container' | 'network', project: string): Map<string, string> {
  const args =
    kind === 'container'
      ? ['container', 'ls', '-a', '--format', `{{.Names}}\t{{.Label "${LABEL_PROJECT}"}}`]
      : ['network', 'ls', '--format', `{{.Name}}\t{{.Label "${LABEL_PROJECT}"}}`];
  const r = spawnSync('docker', args, { encoding: 'utf8' });
  const out = new Map<string, string>();
  if (r.status !== 0) return out;
  for (const line of String(r.stdout ?? '').split('\n')) {
    if (!line.trim()) continue;
    const [name, label] = line.split('\t');
    out.set(name, label ?? '');
  }
  return out;
}

export function findConflicts(config: DevcontainerConfig): Conflict[] {
  if (!dockerAvailable()) return [];
  const plan = plannedComposeNames(config);
  const project = projectId(config);
  const conflicts: Conflict[] = [];

  const existingContainers = listExisting('container', project);
  for (const name of plan.containers) {
    const owner = existingContainers.get(name);
    if (owner === undefined) continue;
    if (owner === project) continue;
    conflicts.push({ kind: 'container', name, owner: owner || '(no label)' });
  }

  const existingNetworks = listExisting('network', project);
  const owner = existingNetworks.get(plan.network);
  if (owner !== undefined && owner !== project) {
    conflicts.push({ kind: 'network', name: plan.network, owner: owner || '(no label)' });
  }

  return conflicts;
}

File: cli/internal/domain/docker_conflicts.go
package domain

import (
	"strings"

	"github.com/joacohbc/my-devcontainer-installer/cli/internal/domain/types"
)

type Conflict struct {
	Kind  string
	Name  string
	Owner string
}

type CaptureFunc func(args []string) (status int, stdout, stderr string)

func FindConflicts(config *types.DevcontainerConfig, dockerAvailable bool, capture CaptureFunc) []Conflict {
	if !dockerAvailable {
		return nil
	}
	containers, network, _, err := PlannedComposeNames(config)
	if err != nil {
		return nil
	}
	project := types.ProjectID(config)

	var conflicts []Conflict

	existingContainers := listExistingDockerResources("container", capture)
	for _, name := range containers {
		owner, found := existingContainers[name]
		if !found {
			continue
		}
		if owner == project {
			continue
		}
		conflicts = append(conflicts, Conflict{Kind: "container", Name: name, Owner: orFallback(owner, "(no label)")})
	}

	existingNetworks := listExistingDockerResources("network", capture)
	if owner, found := existingNetworks[network]; found && owner != project {
		conflicts = append(conflicts, Conflict{Kind: "network", Name: network, Owner: orFallback(owner, "(no label)")})
	}

	return conflicts
}

func listExistingDockerResources(kind string, capture CaptureFunc) map[string]string {
	labelKey := types.LabelProject
	var args []string
	if kind == "container" {
		args = []string{"container", "ls", "-a", "--format", `{{.Names}}\t{{.Label "` + labelKey + `"}}`}
	} else {
		args = []string{"network", "ls", "--format", `{{.Name}}\t{{.Label "` + labelKey + `"}}`}
	}
	_, stdout, _ := capture(args)
	resources := make(map[string]string)
	for _, line := range strings.Split(stdout, "\n") {
		if strings.TrimSpace(line) == "" {
			continue
		}
		parts := strings.SplitN(line, "\t", 2)
		if len(parts) == 2 {
			resources[parts[0]] = parts[1]
		} else {
			resources[parts[0]] = ""
		}
	}
	return resources
}

func orFallback(s, fallback string) string {
	if s == "" {
		return fallback
	}
	return s
}
